File: rk4-rf.c
/*
 * Rabinovich–Fabrikant System
 *
 * Example: ./rk4-rf-dbg 15 NA 1 .01 50000 .05 -.05 .3 .28713 .1
 *          ./rk4-rf-dbg 15 NA 1 .01 50000 .05 -.05 .3 .105 .1 | ./plotPi3d.py
 *
 * (c) 2018-2020 m4r35n357@gmail.com (Ian Smith), for licencing see the LICENCE file
 */

#include <stdlib.h>
#include <assert.h>
#include "taylor-ode.h"

typedef struct {
    real alpha;
    real gamma;
} parameters;

static void *get_p (int argc, char **argv) {
    parameters *p = malloc(sizeof (parameters));
    t_args(argv, argc, &p->alpha, &p->gamma);
    return p;
}

static components ode (real x, real y, real z, void *params) {
    parameters *p = (parameters *)params;
    return (components) {
        .x = y * (z - 1.0 + x * x) + p->gamma * x,
        .y = x * (3.0 * z + 1.0 - x * x) + p->gamma * y,
        .z = - 2.0 * z * (p->alpha + x * y)
    };
}

int main (int argc, char **argv) {
    assert(argc == 11);
    rk4(argc, argv, ode, get_p);
    return 0;
}

File: rk4-rf.py
#!/usr/bin/env python3
#
#  (c) 2018-2023 m4r35n357@gmail.com (Ian Smith), for licencing see the LICENCE file

from sys import argv, stderr
from collections import namedtuple
from ad import Components, Context, rk4

class Parameters(namedtuple('ParametersType', ['α', 'γ'])):
    pass

def get_p():
    return Parameters(α=float(argv[8]), γ=float(argv[9]))

def ode(x, y, z, p):
    a = z + x**2 - 1.0
    b = 4.0 * z - a
    c = p.α + x * y
    return Components(x = y * a + p.γ * x,
                      y = x * b + p.γ * y,
                      z = - 2.0 * z * c)


print(f'RK4: {argv}', file=stderr)
Context.places, _, h, steps = int(argv[1]), int(argv[2]), float(argv[3]), int(argv[4])  # controls
rk4(ode, Context.places, h, steps, float(argv[5]), float(argv[6]), float(argv[7]), get_p())
